File: packages/core/src/workflow/ingestion/benchmarks/gaia/download_gaia_metadata.py
#!/usr/bin/env python3
"""
Download GAIA dataset metadata files
"""

import json
import os
import sys
from pathlib import Path
import requests

def download_gaia():
    """Download GAIA dataset metadata files."""
    
    token = os.environ.get("HF_TOKEN") or os.environ.get("HUGGING_FACE_API_KEY")
    if not token:
        print("Error: No HF_TOKEN or HUGGING_FACE_API_KEY found in environment")
        sys.exit(1)
    
    print("Downloading GAIA dataset metadata...")
    
    # Create output directory
    data_dir = Path(__file__).parent / "output"
    data_dir.mkdir(exist_ok=True)
    
    # Base URL for GAIA dataset files
    base_url = "https://huggingface.co/datasets/gaia-benchmark/GAIA/resolve/main/2023"
    
    # Files to download
    files = {
        "validation": "validation/metadata.jsonl",
        "test": "test/metadata.jsonl"
    }
    
    headers = {
        "Authorization": f"Bearer {token}"
    }
    
    for split_name, filepath in files.items():
        print(f"Downloading {split_name} metadata...")
        
        url = f"{base_url}/{filepath}"
        
        try:
            response = requests.get(url, headers=headers)
            response.raise_for_status()
            
            # Parse JSONL and convert to JSON
            items = []
            for line in response.text.strip().split('\n'):
                if line:
                    try:
                        item = json.loads(line)
                        # Convert to our format
                        json_item = {
                            "task_id": item.get("task_id"),
                            "Question": item.get("Question"),
                            "Level": int(item.get("Level", 0)),
                        }
                        
                        # Add optional fields
                        if "Final answer" in item and item["Final answer"]:
                            json_item["Final answer"] = str(item["Final answer"])
                        if "file_name" in item and item["file_name"]:
                            json_item["file_name"] = str(item["file_name"])
                        
                        items.append(json_item)
                    except json.JSONDecodeError as e:
                        print(f"Error parsing line: {e}")
                        print(f"Line content: {line[:100]}...")
            
            # Save to JSON file
            output_file = data_dir / f"{split_name}.json"
            with open(output_file, "w", encoding="utf-8") as f:
                json.dump(items, f, indent=2, ensure_ascii=False)
            
            print(f"Saved {len(items)} items to {output_file}")
            
        except requests.exceptions.HTTPError as e:
            print(f"Error downloading {split_name}: HTTP {e.response.status_code}")
            print(f"Response: {e.response.text[:200]}...")
        except Exception as e:
            print(f"Error processing {split_name}: {e}")
            import traceback
            traceback.print_exc()
    
    # Create metadata file
    metadata = {
        "dataset": "gaia-benchmark/GAIA",
        "config": "2023",
        "download_method": "metadata"
    }
    
    with open(data_dir / "metadata.json", "w") as f:
        json.dump(metadata, f, indent=2)
    
    print("\nDownload complete!")
    print(f"Data saved to: {data_dir}")

if __name__ == "__main__":
    download_gaia()

File: packages/core/src/evaluation/benchmarks/gaia/download_gaia_metadata.ts
import { mkdirSync, writeFileSync } from "node:fs"
import { dirname, join } from "node:path"
import { fileURLToPath } from "node:url"
import { envi } from "@core/utils/env.mjs"

type GaiaRow = {
  task_id: string
  Question: string
  Level: number
  "Final answer"?: string
  file_name?: string
}

const BASE_FILES_URL = "https://huggingface.co/datasets/gaia-benchmark/GAIA/resolve/main/2023"

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)
const OUTPUT_DIR = join(__dirname, "output")

function getAuthToken(): string {
  if (!envi.HF_TOKEN) {
    throw new Error("HF_TOKEN is not set in environment variables")
  }
  return envi.HF_TOKEN
}

async function fetchText(url: string, authToken?: string): Promise<string> {
  const headers: HeadersInit = {}
  if (authToken) headers.Authorization = `Bearer ${authToken}`
  const res = await fetch(url, { headers })
  if (!res.ok) {
    const body = await res.text().catch(() => "")
    throw new Error(`HTTP ${res.status} ${res.statusText}${body ? ` - ${body}` : ""}`)
  }
  return res.text()
}

function parseJsonlToArray(text: string): GaiaRow[] {
  const rows: GaiaRow[] = []
  for (const line of text.split("\n")) {
    const trimmed = line.trim()
    if (!trimmed) continue
    try {
      const item = JSON.parse(trimmed)
      const mapped: GaiaRow = {
        task_id: item.task_id,
        Question: item.Question,
        Level: Number(item.Level ?? 0),
      }
      if (item["Final answer"]) mapped["Final answer"] = String(item["Final answer"]) // keep key
      if (item.file_name) mapped.file_name = String(item.file_name)
      // Skip instances with files to align with loader default
      if (!mapped.file_name && mapped.task_id !== "0-0-0-0-0") rows.push(mapped)
    } catch {
      // skip malformed lines
    }
  }
  return rows
}

function ensureOutputDir(): void {
  mkdirSync(OUTPUT_DIR, { recursive: true })
}

function saveJSON(filename: string, data: unknown): void {
  const filepath = join(OUTPUT_DIR, filename)
  writeFileSync(filepath, JSON.stringify(data, null, 2), { encoding: "utf-8" })
}

export async function downloadGAIAMetadata(): Promise<void> {
  const token = getAuthToken()
  ensureOutputDir()

  const files: Record<string, string> = {
    validation: "validation/metadata.jsonl",
    test: "test/metadata.jsonl",
  }

  const results: Record<"validation" | "test", GaiaRow[]> = {
    validation: [],
    test: [],
  }

  for (const [split, file] of Object.entries(files)) {
    try {
      const text = await fetchText(`${BASE_FILES_URL}/${file}`, token)
      const rows = parseJsonlToArray(text)
      results[split as keyof typeof results] = rows
      saveJSON(`${split}.json`, rows)
    } catch (e: any) {
      // eslint-disable-next-line no-console
      console.warn(`Failed to download ${split} metadata:`, e?.message || e)
    }
  }

  saveJSON("metadata.json", {
    dataset: "gaia-benchmark/GAIA",
    config: "2023",
    download_method: "metadata",
    splits: ["validation", "test"],
    total_items: results.validation.length + results.test.length,
  })

  // eslint-disable-next-line no-console
  console.log(`GAIA metadata downloaded to ${OUTPUT_DIR}`)
}

if (import.meta.url === `file://${__filename}`) {
  downloadGAIAMetadata().catch(() => process.exit(1))
}
